src/System.Diagnostics.Process/src/System/Diagnostics/ProcessWaitState.Unix.cs

```diff
@@ -0,0 +1,503 @@
+
+    /// <summary>Exit information and waiting capabilities for a process.</summary>
+    internal sealed class ProcessWaitState : IDisposable
+    {
+        /// <summary>
+        /// Finalizable holder for a process wait state. Instantiating one
+        /// will ensure that a wait state object exists for a process, will
+        /// grab it, and will increment its ref count.  Dropping or disposing
+        /// one will decrement the ref count and clean up after it if the ref
+        /// count hits zero.
+        /// </summary>
+        internal sealed class Holder : IDisposable
+        {
+            internal ProcessWaitState _state;
+
+            internal Holder(int processId)
+            {
+                _state = ProcessWaitState.AddRef(processId);
+            }
+
+            ~Holder()
+            {
+                if (_state != null)
+                {
+                    _state.ReleaseRef();
+                }
+            }
+
+            public void Dispose()
+            {
+                if (_state != null)
+                {
+                    _state.ReleaseRef();
+                    _state = null;
+                    GC.SuppressFinalize(this);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Global table that maps process IDs to the associated shared wait state information.
+        /// </summary>
+        private static readonly Dictionary<int, ProcessWaitState> s_processWaitStates =
+            new Dictionary<int, ProcessWaitState>();
+
+        /// <summary>
+        /// Ensures that the mapping table contains an entry for the process ID,
+        /// increments its ref count, and returns it.
+        /// </summary>
+        /// <param name="processId">The process ID for which we need wait state.</param>
+        /// <returns>The wait state object.</returns>
+        internal static ProcessWaitState AddRef(int processId)
+        {
+            lock (s_processWaitStates)
+            {
+                ProcessWaitState pws;
+                if (!s_processWaitStates.TryGetValue(processId, out pws))
+                {
+                    pws = new ProcessWaitState(processId);
+                    s_processWaitStates.Add(processId, pws);
+                }
+                pws._outstandingRefCount++;
+                return pws;
+            }
+        }
+
+        /// <summary>
+        /// Decrements the ref count on the wait state object, and if it's the last one,
+        /// removes it from the table.
+        /// </summary>
+        internal void ReleaseRef()
+        {
+            lock (ProcessWaitState.s_processWaitStates)
+            {
+                ProcessWaitState pws;
+                bool foundState = ProcessWaitState.s_processWaitStates.TryGetValue(_processId, out pws);
+                Debug.Assert(foundState);
+                if (foundState)
+                {
+                    if (--pws._outstandingRefCount == 0)
```

Comment:
Would you mind doing the decrement outside the test?  I know they are isomorphic, but I think it reads better.